### ISSUE TITLE
V6 upgrade

### ISSUE THREAD

**Issue by danielmarbach** [Particular/ServiceControl#623]
@mikeminutillo @johnsimons @gbiellem 

I would like to take a stab at this with your help. I'd be OK to start earlier in the morning to create more overlapp with you so that we can do a few pair programming sessions or whatever is required to get feedback from you. Any ideas, suggestions to pull this off?

**Comment by gbiellem** [comment]
@danielmarbach sounds good

I don't think we should tag it as part of V6 though as I don't think it's a blocker for the V6 release.
If this isn't the case it would be good if you can detail why you think it is.

**Comment by johnsimons** [comment]
:+1: on what @gbiellem said

**Comment by danielmarbach** [comment]
I've expressed my opinion about this several times. Let's jump on a call

**Comment by gbiellem** [comment]
> Let's jump on a call

@danielmarbach  sure. It'll have to be next week. I've quit for the day,  

> I've expressed my opinion about this several times.

Great but i was hoping to capture your reasoning here so we don't have to go through this again when we are queried about why it's part of V6 and why we are spending time on it.  I've not been in any conversation about it so I can't speak for you

**Comment by danielmarbach** [comment]
I even stated it in GH issue (which I can't find now...) that in my opinion
v6 is not done before we dogfooded it with SC.

**Comment by mikeminutillo** [comment]
I think we could spike this to ensure that the seam is correct. There is a non-trivial amount of effort involved in doing releasing a version of ServiceControl on NSB v6:

- Updating to v6 requires us to update the RavenDB Persistence as well which will require an update to Raven DB version 3. This will require a data conversion.
- ServiceControl (rightly or wrongly) subscribes to it's own events. We did this as a hack in v4 and v5. We do not know if it will continue to work in v6.
- The ingestion process is also based on hacking the v5 pipeline. This infrastructure is different in v6 and will need to be tested. 
- The transport seams (which the retry mechanism relies upon) have changed in v6.
- The update to v5 was specifically to support a feature in the SQL Transport (multiple databases) which is being dropped from v6. As we only maintain a single version of ServiceControl, moving forward would also mean dropping this feature.

If the decision has been made to hold the release of NSB v6 until all affected downstream repos have been updated then ServiceControl is one of those repos and needs to be updated. If we don't do that, then we run the risk that the seam is not right and updating ServiceControl is difficult/impossible. This could lead to a situation where v6 releases new features that are not compatible with ServiceControl and we can't quickly respond. That is undesirable.

Everything NServiceBus should be wire compatible. ServiceControl does 3 things:
- Audit Ingestion - is reasonably simple. As long as endpoints put messages in the right queue, we will load them and persist them in the database. It is unlikely that a change in the transport that will affect this.
- Heartbeats / Custom Checks - is controlled on both ends by us. We own the plugin that goes into the client so we can control what goes on to the wire. 
- Error ingestion and recoverability - is where it gets messy. This is where we have to be able to forward messages back into the live system so they better be compatible with users endpoints. Having said that, there are a few headers that we rely upon for routing information and if those are still valid we are doing a copy-paste of the message (headers and all) from the error queue to back to the input queue of the processing endpoint.

The best bet is probably for us to spike a solution on v6 to test a few things and stop without releasing it to customers. We should also ensure that v6 endpoints can communicate correctly with the version of ServiceControl available today. This is arguably *more* important.

**Comment by danielmarbach** [comment]
:+1:  what @mikeminutillo said

**Comment by mikeminutillo** [comment]
So next steps. @danielmarbach will create a new PR targeting v6 and upgrade ServiceControl to target the latest unstable v6. This will require handler signature migration at the very least to get it to build. Until this is building and tests are reliably passing, @gbiellem will mute the build for this PR on the build server. 

Once this is done we will look at the areas which are most likely to break (ingestion, migration, self-subscription, recoverability) to decide what items we want to tackle and how much time we want to dedicate to it.

**Comment by mikeminutillo** [comment]
One other thing. As we intend to start pulling things into a separate logical context in ServiceControl we could go an extra step and pull them into separate physical contexts and base those new contexts on v6. Heartbeats and Custom Checks is supposed to be the first off the rack for this and requires minimal migration. 

This actually makes a lot of sense as each could have it's own circuit breakers and fail independently of each other.

**Comment by gbiellem** [comment]
> This actually makes a lot of sense as each could have it's own circuit breakers and fail independently of each other. 

This would make me very happy

**Comment by mikeminutillo** [comment]
Good. Because the Management Utility will need to control multiple
processes when we make this change :)

**Comment by mikeminutillo** [comment]
As discussed with @danielmarbach, we can still explore the migration to v6 to ensure that the seam is right in the short term.

**Comment by danielmarbach** [comment]
@mat-mcloughlin joins this initiative as well. We are doing it today together with pair programming to spread the knowledge.

**Comment by danielmarbach** [comment]
Link the following PR

https://github.com/Particular/ServiceControl/pull/627

Status update on the PR

**Comment by gbiellem** [comment]
@danielmarbach  seems you've wrapped up the spike .  Any reason to keep this open?

**Comment by boblangley** [comment]
New Azure ServiceBus Transport features cannot be supported until ServiceControl supports v6 (and can be upgraded to >= v7 of ASB). 1 feature raised in a recent support case includes setting the [bus connectivity mode](https://msdn.microsoft.com/en-us/library/microsoft.servicebus.connectivitysettings.mode.aspx) options for the transport which was introduced in [this commit](https://github.com/Particular/NServiceBus.AzureServiceBus/commit/94f000a7c5e95e016ae8adfe428d7dec3c01bd15).
